src/interpreter/InterpreterNode.h

```diff
@@ -112,16 +111,142 @@ class InterpreterNode {
         return shadow;
     }
 
+    /** @brief get relation from handle */
+    InterpreterRelation* getRelation() const {
+        assert(relHandle && "No relation cached\n");
+        return (*relHandle).get();
+    }
+
+protected:
+    enum InterpreterNodeType type;
+    const RamNode* shadow;
+    RelationHandle* const relHandle;
+};
+
+/**
+ * @class InterpreterCompundNode
+ * @brief Compound node represents interpreter node with a list of children.
+ */
+class InterpreterCompundNode : public InterpreterNode {
+    using NodePtrVec = std::vector<std::unique_ptr<InterpreterNode>>;
+
+public:
+    InterpreterCompundNode(enum InterpreterNodeType ty, const RamNode* sdw, NodePtrVec children = {},
+            RelationHandle* relHandle = nullptr)
+            : InterpreterNode(ty, sdw, relHandle), children(std::move(children)) {}
+
     /** @brief get children of node */
     inline const InterpreterNode* getChild(std::size_t i) const {
         return children[i].get();
     }
 
-    /** @brief get data */
-    inline size_t getData(std::size_t i) const {
-        return data[i];
+    /** @brief get list of all children */
+    const NodePtrVec& getChildren() const {
+        return children;
+    }
+
+protected:
+    NodePtrVec children;
+};
+
+/**
+ * @class InterpreterUnaryNode
+ * @brief Unary node represents interpreter node with a single child
+ */
+class InterpreterUnaryNode : public InterpreterNode {
+public:
+    InterpreterUnaryNode(enum InterpreterNodeType ty, const RamNode* sdw,
+            std::unique_ptr<InterpreterNode> child, RelationHandle* relHandle = nullptr)
+            : InterpreterNode(ty, sdw, relHandle), child(std::move(child)) {}
+
+    inline const InterpreterNode* getChild() const {
+        return child.get();
+    }
+
+protected:
+    std::unique_ptr<InterpreterNode> child;
+};
+
+/**
+ * @class InterpreterBinaryNode
+ * @brief Binary node represents interpreter node with two children.
+ */
+class InterpreterBinaryNode : public InterpreterNode {
+public:
+    InterpreterBinaryNode(enum InterpreterNodeType ty, const RamNode* sdw,
+            std::unique_ptr<InterpreterNode> lhs, std::unique_ptr<InterpreterNode> rhs,
+            RelationHandle* relHandle = nullptr)
+            : InterpreterNode(ty, sdw, relHandle), lhs(std::move(lhs)), rhs(std::move(rhs)) {}
+
+    /** @brief get left child of node */
+    inline const InterpreterNode* getLhs() const {
+        return lhs.get();
+    }
+
+    /** @brief get right child of node */
+    inline const InterpreterNode* getRhs() const {
+        return rhs.get();
     }
 
+protected:
+    std::unique_ptr<InterpreterNode> lhs;
+    std::unique_ptr<InterpreterNode> rhs;
+};
+
+/**
+ * @class InterpreterSuperInst
+ * @brief This class encodes information for a super-instruction, which is
+ *        used to eliminate Number and TupleElement in index/project/existence operation.
+ */
+class InterpreterSuperInst {
```

Comment:
`Inst` can be short for many things, including related things. Use the full word for clarity:
```suggestion
class InterpreterSuperInstruction {
```